### PR TITLE
Enable multiple secrets dirs

### LINKED ISSUE #3: Support more than one directory for settings management
very similar to https://github.com/pydantic/pydantic/issues/2284

I would love to be able to leverage built-in settings configuration, but in order to do that, I need support for both ConfigMaps and Secrets

i don't know if a file per setting is absolutely necessary, but it would be great if instead of a `secrets_dir: str|None` you had a `directories: List[str]|None` or `search_dirs: List[str]|None` argument, where settings were loaded by iterating over 1..N directories using much the same logic as today.

beyond just allowing use of both ConfigMaps and Secrets, this _could_ then produce conflicts, which instead of a defect, would actually be a very handy a feature, it allows for overrides (such as when running locally) this can be accomplished by having the expectation that the _last found match wins_. 

```
class Settings(BaseSettings):
    my_secret_data: str
    my_not_secret_data: str
    my_overridden_data: str


    class Config:
        search_dirs = [ '/run/config' ,   '/run/secrets', ]
```

with files
```
/run/config/my_not_secret_data # foo
/run/config/my_overridden_data # bar
/run/secrets/my_overridden_data # baz
/run/secrets/my_secret_data # fu
```
pragmatic expectation:
all 4 files get opened/read/parsed/validated

resulting object is

```
s = Settings() # s.my_secret_data == 'fu', s.my_not_secret_data == 'foo', s.my_overridden_data == 'baz', 
```

### PR DESCRIPTION
**Pull request by makukha**
* Allow `secrets_dir` to be a tuple or list of directories

Closes:
* #3
* #371

### CODE REVIEW

**Comment by makukha** [comment]
Ready for review!

**Comment by makukha** [comment]
@hramezani it seems that I don't have permissions to ask for review, but this PR is probably ready.

**Comment by hramezani** [comment]
> @hramezani it seems that I don't have permissions to ask for review, but this PR is probably ready.

Thanks @makukha for the PR. I will review it later

File: tests/test_settings.py [hunk 1]
```diff
@@ -1416,6 +1416,30 @@ class Settings(BaseSettings):
...
+        foo2: str
+        foo3: str
+
+        model_config = SettingsConfigDict(secrets_dir=(d1, d2))
+
+    assert Settings().model_dump() == {
+        'foo1': 'foo1_dir1_secret_value_str',
+        'foo2': 'foo2_dir2_secret_value_str',  # dir2 takes priority
+        'foo3': 'foo3_dir2_secret_value_str',
+    }
```
Comment (hramezani) [inline tests/test_settings.py:1440]: please add another model with reverse `secrets_dir`

File: pydantic_settings/sources.py [hunk 2]
```diff
@@ -594,14 +594,17 @@ def __call__(self) -> dict[str, Any]:
...
+        for path in self.secrets_paths:
+            if not path.exists():
+                warnings.warn(f'directory "{path}" does not exist')
+                return secrets
 
-        if not self.secrets_path.is_dir():
-            raise SettingsError(f'secrets_dir must reference a directory, not a {path_type_label(self.secrets_path)}')
+        for path in self.secrets_paths:
+            if not path.is_dir():
+                raise SettingsError(f'secrets_dir must reference a directory, not a {path_type_label(path)}')
```
Comment (hramezani) [inline pydantic_settings/sources.py:607]: There is a test for this error that tests single dir. please add a test for multiple directories as well.

File: pydantic_settings/sources.py [hunk 3]
```diff
@@ -594,14 +594,17 @@ def __call__(self) -> dict[str, Any]:
...
-        self.secrets_path = Path(self.secrets_dir).expanduser()
+        secrets_dirs = [self.secrets_dir] if isinstance(self.secrets_dir, (str, os.PathLike)) else self.secrets_dir
+        self.secrets_paths = [Path(p).expanduser() for p in secrets_dirs]
 
-        if not self.secrets_path.exists():
-            warnings.warn(f'directory "{self.secrets_path}" does not exist')
-            return secrets
+        for path in self.secrets_paths:
+            if not path.exists():
+                warnings.warn(f'directory "{path}" does not exist')
```
Comment (hramezani) [inline pydantic_settings/sources.py:602]: There is a test for this warning that tests single dir. please add a test for multiple directories as well.

File: pydantic_settings/sources.py [hunk 4]
```diff
@@ -639,18 +642,20 @@ def get_field_value(self, field: FieldInfo, field_name: str) -> tuple[Any, str,
...
-                return path.read_text().strip(), field_key, value_is_complex
-            else:
-                warnings.warn(
-                    f'attempted to load secret file "{path}" but found a {path_type_label(path)} instead.',
-                    stacklevel=4,
-                )
+                if path.is_file():
+                    return path.read_text().strip(), field_key, value_is_complex
+                else:
+                    warnings.warn(
```
Comment (hramezani) [inline pydantic_settings/sources.py:655]: There is a test for this warning that tests single dir. please add a test for multiple directories as well.

**Comment by makukha** [comment]
@hramezani adding tests for warnings and errors was a good idea, fixed a bug. Also, I updated docs describing behaviour when secrets dirs can be missing or can be files.

File: tests/test_settings.py [hunk 5]
```diff
@@ -1544,6 +1593,37 @@ class Settings(BaseSettings):
...
+
+@pytest.mark.skipif(sys.platform.startswith('win'), reason='windows paths break regex')
+def test_secrets_missing_location_multiple_one(tmp_path):
+    class Settings(BaseSettings):
+        foo: Optional[str] = None
+
+    (d1 := tmp_path / 'dir1').mkdir()
+    (d1 / 'foo').write_text('secret_value')
+
+    with pytest.warns() as record:
```
Comment (hramezani) [inline tests/test_settings.py:1618]: can't you use `with pytest.warns(UserWarning, match=...)` for this test and the others?

Comment (makukha) [reply tests/test_settings.py]: Fixed for two cases where a single warning is emitted. For other cases, we have multiple warnings, and need a `record` object from `pytest.warns() as record` to check both. I wouldn't further simplify those cases.

**Review by hramezani — APPROVED** [review]

**Comment by hramezani** [comment]
Thanks @makukha

**Comment by makukha** [comment]
@hramezani thank you for supporting this awesome package!

**Pull request merged**